tensordict/prototype/tensorclass.py

```diff
@@ -158,56 +159,75 @@ def __torch_function__(
-
-            init(self, **new_kwargs)
-            self.tensordict = TensorDict(
-                {
-                    key: _get_typed_value(value)
-                    for key, value in kwargs.items()
-                    if key not in ("batch_size",)
-                },
-                batch_size=batch_size,
-                device=device,
+    @functools.wraps(init)
+    def wrapper(self, *args, batch_size, device=None, **kwargs):
+        for value, key in zip(args, self.__dataclass_fields__):
+            if key in kwargs:
+                raise ValueError(f"The key {key} is already set in kwargs")
+            kwargs[key] = value
+
+        for key, field in self.__dataclass_fields__.items():
+            if field.default_factory is not dataclasses.MISSING:
+                default = field.default_factory()
+            else:
+                default = field.default
+            if default not in (None, dataclasses.MISSING):
+                kwargs.setdefault(key, default)
+
+        missing_params = [p for p in required_params if p not in kwargs]
+        if missing_params:
+            n_missing = len(missing_params)
+            raise TypeError(
+                f"{self.__class__.__name__}.__init__() missing {n_missing} "
+                f"required positional argument{'' if n_missing == 1 else 's'}: "
+                f"""{", ".join(f"'{name}'" for name in missing_params)}"""
             )
 
+        self.tensordict = TensorDict(
+            {}, batch_size=batch_size, device=device, _run_checks=False
+        )
+        init(self, **{key: _get_typed_value(value) for key, value in kwargs.items()})
+
+    new_params = [
+        inspect.Parameter("batch_size", inspect.Parameter.KEYWORD_ONLY),
+        inspect.Parameter("device", inspect.Parameter.KEYWORD_ONLY, default=None),
+    ]
+    wrapper.__signature__ = init_sig.replace(parameters=params + new_params)
+
     return wrapper
 
 
-def _build_from_tensordict(cls, tensordict):
-    return cls(_tensordict=tensordict)
+def _from_tensordict_wrapper(expected_keys):
+    def wrapper(cls, tensordict):
+        if not all(key in expected_keys for key in tensordict.keys()):
+            raise ValueError(
+                f"Keys from the tensordict ({set(tensordict.keys())}) must "
+                f"correspond to the class attributes ({expected_keys})."
+            )
+        # bypass initialisation. this means we don't incur any overhead creating an
+        # empty tensordict and writing values to it. we can skip this because we already
+        # have a tensordict to use as the underlying tensordict
+        tc = cls.__new__(cls)
+        tc.__dict__["tensordict"] = tensordict
+        # since we aren't calling the dataclass init method, we need to manually check
+        # whether a __post_init__ method has been defined and invoke it if so
+        if hasattr(tc, "__post_init__"):
+            tc.__post_init__()
+        return tc
```

Comment:
I added this to reduce overhead when constructing from `TensorDict`. We bypass the regular constructor which means we don't create a tensordict or have to set the attributes. The only thing we need to make sure we do is run the `__post_init__` method if it exists because we're no longer invoking the dataclass' init method.

I think this is ok, but if you can think of any edge cases I might have missed please let me know!

See relevant discussion here: https://github.com/pytorch-labs/tensordict/pull/175#discussion_r1084278607

Reply:
Happy with that solution